pkg/runner/test_runner.go

```diff
@@ -1,287 +1,750 @@
+	tr.log.Debugf("step states:")
+	for i, ss := range tr.steps {
+		tr.log.Debugf("  %d %s %t %t %v", i, ss, ss.stepRunning, ss.readerRunning, ss.runErr)
+	}
+
+	// Is there a useful error to report?
+	if runErr != nil {
+		return nil, runErr
+	}
+
+	// Has the run been canceled?
+	select {
+	case <-ctx.Done():
+		return nil, statectx.ErrCanceled
+	default:
+	}
 
-	errCh := make(chan error, 1)
+	// Have we been asked to pause? If yes, is it safe to do so?
+	select {
+	case <-ctx.Paused():
+		if !resumeOk {
+			tr.log.Warningf("paused but not ok to resume")
+			break
+		}
+		rs := &resumeStateStruct{
+			Version: resumeStateStructVersion,
+			JobID:   jobID, RunID: runID,
+			Targets: tr.targets,
+		}
+		resumeState, runErr = json.Marshal(rs)
+		if runErr != nil {
+			tr.log.Errorf("unable to serialize the state: %s", runErr)
+		} else {
+			runErr = statectx.ErrPaused
+		}
+	default:
+		// We are not pausing and yet some targets were left in flight.
+		if len(inFlightTargets) > 0 {
+			ts := inFlightTargets[0]
+			runErr = &cerrors.ErrTestStepLostTargets{
+				StepName: tr.steps[ts.CurStep].sb.TestStepLabel,
+				Target:   ts.tgt.ID,
+			}
+		}
+	}
+	return resumeState, runErr
+}
+
+func (tr *testRunner) waitStepRunners(ctx statectx.Context) error {
+	tr.log.Debugf("waiting for step runners to finish")
+	swch := make(chan struct{})
 	go func() {
-		log.Infof("running pipeline")
-		errCh <- testPipeline.run(completedTargets)
+		tr.mu.Lock()
+		defer tr.mu.Unlock()
+		for {
+			ok := true
+			for _, ss := range tr.steps {
+				// numRunning == 1 is also acceptable: we allow the Run() goroutine
+				// to continue in case of error, if the result processor decided
+				// to abandon its runner, there's nothing we can do.
+				switch {
+				case !ss.stepRunning && !ss.readerRunning:
+				// Done
+				case ss.stepRunning && ss.readerRunning:
+					// Still active
+					ok = false
+				case !ss.stepRunning && ss.readerRunning:
+					// Transient state, let it finish
+					ok = false
+				case ss.stepRunning && !ss.readerRunning:
+					// This is possible if plugin got stuck and result processor gave up on it.
+					// If so, it should have left an error.
+					if ss.runErr == nil {
+						tr.log.Errorf("%s: result processor left runner with no error", ss)
+						// There's nothing we can do at this point, fall through.
+					}
+				}
+			}
+			if ok {
+				close(swch)
+				return
+			}
+			tr.cond.Wait()
+		}
 	}()
+	tr.cond.Signal()
+	select {
+	case <-swch:
+		tr.log.Debugf("step runners finished")
+		tr.mu.Lock()
+		defer tr.mu.Unlock()
+		return tr.checkStepRunners()
+	case <-time.After(tr.shutdownTimeout):
+		tr.log.Errorf("step runners failed to shut down correctly")
+		// If there is a step with an error set, use that.
+		if err := tr.checkStepRunners(); err != nil {
+			return err
+		}
+		// If there isn't, enumerate ones that were still running at the time.
+		err := &cerrors.ErrTestStepsNeverReturned{}
+		tr.mu.Lock()
```

Comment:
Because we acquire the lock in the goroutine above, we get blocked here until that goroutine finishes (swch channel is closed)

Reply:
not quite. cond.Wait() releases the lock while waiting.

Reply:
Sorry, it seems to be ok because of tr.cond.Wait() will release the mutex